volatility3/framework/layers/arm.py

```diff
@@ -0,0 +1,475 @@
+            ) as excp:
+                """
+                Contiguous pages might not be mapped, but if we specifically ignore errors, we still want to read the n + 1 page as it might be mapped.
+                """
+                if not ignore_errors:
+                    raise
+                # We can jump more if we know where the page fault failed
+                if isinstance(excp, exceptions.PagedInvalidAddressException):
+                    mask = (1 << excp.invalid_bits) - 1
+                else:
+                    mask = (
+                        1
+                        << (
+                            self._ttbs_granules[self._virtual_addr_space].bit_length()
+                            - 1
+                        )
+                    ) - 1
+                length_diff = mask + 1 - (offset & mask)
+                length -= length_diff
+                offset += length_diff
+            else:
+                yield offset, chunk_size, chunk_offset, chunk_size, layer_name
+                length -= chunk_size
+                offset += chunk_size
+
+    def _get_virtual_addr_ranges(
+        self,
+    ) -> List[Tuple[int]]:
+        """Returns the virtual address space ranges as [(LOW_START, LOW_END), (HIGH_START, HIGH_END)]"""
+        # [2], see source/arch/arm64/include/asm/memory.h#L62
+        ttb0_start = 0
+        ttb0_size = 1 << (self._ttbs_bitsizes[0] - 1)
+        ttb0_end = ttb0_start + (ttb0_size - 1)
+        ttb1_end = 2**64 - 1
+        ttb1_size = 1 << (self._ttbs_bitsizes[1] - 1)
+        ttb1_start = ttb1_end - (ttb1_size - 1)
+
+        return [(ttb0_start, ttb0_end), (ttb1_start, ttb1_end)]
+
+    def is_valid(self, offset: int, length: int = 1) -> bool:
+        """Returns whether the address offset can be translated to a valid
+        address."""
+        try:
+            # TODO: Consider reimplementing this, since calls to mapping can call is_valid
+            return all(
+                [
+                    self._context.layers[layer].is_valid(mapped_offset)
+                    for _, _, mapped_offset, _, layer in self.mapping(offset, length)
+                ]
+            )
+        except exceptions.InvalidAddressException:
+            return False
+
+    def is_dirty(self, offset: int) -> bool:
+        """Returns whether the page at offset is marked dirty"""
+        return self._page_is_dirty(self._translate_entry(offset)[2])
+
+    @staticmethod
+    def _page_is_dirty(entry: int) -> bool:
+        """Returns whether a particular page is dirty based on its entry."""
+        # [2], see arch/arm64/include/asm/pgtable-prot.h#L18
+        return bool(entry & (1 << 55))
+
+    @classproperty
+    @functools.lru_cache()
+    def minimum_address(cls) -> int:
+        return 0
+
+    @classproperty
+    @functools.lru_cache()
+    def maximum_address(cls) -> int:
+        return (1 << cls._maxvirtaddr) - 1
+
+    def __canonicalize(self, addr: int) -> int:
```

Comment:
These two don't appear to get called from anywhere and are strictly private?  Are they necessary?  If they can't be depended on in children of this class, then they don't seem all that useful?

Reply:
Inspired from : https://github.com/volatilityfoundation/volatility3/blob/develop/volatility3/framework/layers/intel.py#L93 and https://github.com/volatilityfoundation/volatility3/blob/develop/volatility3/framework/layers/intel.py#L119.  

I realize now that I messed up the visibility of canonicalize, and will fix it. I did not investigated which higher level code was likely to use these functions, but the Intel one have them so they aren't useless (even if I think `maximum_address` is just a convenient internal getter).